Add docs to oz docsite (#298)
OpenZeppelin/nile: docs/modules/ROOT/pages/index.adoc

:pyenv: link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.4.0b/src/openzeppelin/access/ownable/library.cairo[pyenv]

= Overview

Nile is a CLI tool to develop or interact with StarkNet projects written in Cairo. It consists of different components for developing, compiling, testing, and deploying your smart contracts and dApps, providing a CLI for executing tasks, and a Runtime Environment (NRE) for scripting.

The package is designed to be extensible and very customizable by using plugins. This guide will take you through the installation of our recommended setup, but as we expect a lot of functionality coming from plugins, you are free to customize it.

> TIP: You can get a list of plugins extending Nile functionality from the xref:plugins.adoc[Plugins] section.

Reviewer: i wouldn't advertise this list aggressively, as i don't think it is very big. is there anything else we'd rather highlight in here?

Author: Good point. I can't think of anything else atm though. I removed it for now at least.

:oz-account: link:https://github.com/OpenZeppelin/cairo-contracts/blob/main/src/openzeppelin/account/presets/Account.cairo[Openzeppelin Account]

== Quickstart

> Using a sample contract, we will explore the basics of creating a Nile project. We will test it locally, deploy an account and the contract to a devnet node, and send transactions through the account to the contract.

Reviewer: piggybacking on [Andrew's comments](https://github.com/OpenZeppelin/nile/pull/298#discussion_r1031075746), here we mention the node but we never show how to use it.

i'll mix that with the following suggestion:

<img width="208" alt="image" src="https://user-images.githubusercontent.com/1703831/209575694-9d5af8d4-0400-484e-8cb3-fe2913d2729e.png">

I think everything below `Compiling` should be a subsection of a wider "Interacting with a node" or "with StarkNet" section. and i would add an additional subsection for scripting, something like "Custom scripts" or "Custom tasks"

Author: I added an "Interacting with Starknet" section including running a local node, the only caveat I find is that we lose the index entries (in the section to the right).

For the custom scripts not sure what to add, when Quickstart's next page is NRE and scripting already.

=== Requirements

==== GMP for fastecdsa

Before installing Cairo on your machine, you need to install `gmp`:

[,bash]
----
sudo apt install -y libgmp3-dev # linux
brew install gmp # mac
----

TIP: If you have any trouble installing gmp on your Apple M1 computer, https://github.com/OpenZeppelin/nile/issues/22[here's a list of potential solutions].

==== Supported Python versions

Some Nile dependencies have specific python version requirements. Therefore we recommend using a python version manager like {pyenv}, and virtual environments to avoid conflicts.

IMPORTANT: Currently supported Python versions are >=3.8 and <3.10.

=== Installation

Create a folder for your project and cd into it:

[,sh]
----
mkdir myproject && cd myproject
----

Create a virtualenv and activate it:

[,sh]
----
python3 -m venv env
source env/bin/activate
----

IMPORTANT: Review xref:index.adoc#requirements[Requirements]

Install Nile:

[,sh]
----

> (env): pip install cairo-nile

Reviewer: ```suggestion
pip install cairo-nile
```
not sure about adding shell prefixes to these instructions

Author: I think it could help the users to avoid making the mistake of not being inside the env. I also think it could add clarity to add a dollar $ sign before commands to differentiate it from the output, like:

<img width="789" alt="Screenshot 2022-12-27 at 17 13 39" src="https://user-images.githubusercontent.com/26070224/209692555-d45e4ad4-a319-4d86-a5ea-e651761fe414.png">

Reviewer: i think the `$` or `(env)` should only be there when we're simply showing how it looks. when it comes to instructions/steps to follow, i favor the plain command because it's easier to copy and paste. the `$` hinders the experience of using this button:

<img width="54" alt="image" src="https://user-images.githubusercontent.com/1703831/209766655-f7ae421f-58d9-468c-b213-2709df4190e2.png">

Author: That's a great point. Following up on that, should we remove then the output from the code boxes? And what about the code boxes with multiple commands or examples for mac and linux?

```
(env): nile compile
🤖 Compiling all Cairo contracts in the contracts directory
🔨 Compiling contracts/contract.cairo
✅ Done
```

```
sudo apt install -y libgmp3-dev # linux
brew install gmp # mac
```

```
python3 -m venv env
source env/bin/activate
```

Reviewer: multiple command ones should work as they are. the mac and linux is one an acceptable exception, but the answer for all cases is that it they can be split up:

```sh
nile compile
```
```sh
🤖 Compiling all Cairo contracts in the contracts directory
🔨 Compiling contracts/contract.cairo
✅ Done
```
------
```sh
sudo apt install -y libgmp3-dev # linux
```
```sh
brew install gmp # mac
```

----

Use `nile init` to quickly set up your development environment:

[,sh]
----
(env): nile init
🗄 Creating project directory tree
⛵️ Nile project ready! Try running:

nile compile
----

=== Testing

`nile init` creates a sample Cairo contract and test for you. Check *contracts/contract.cairo* and *tests/test_contract.py* for the source code.

Run `pytest` to run the test suite against the Smart Contracts:

[,sh]
----
(env): pytest tests/
----

> TIP: For a more in-deep guide on testing with parallelism and coverage, check our xref:testing.adoc[Testing] guide.

Reviewer: ```suggestion
TIP: For a more in-deep guide on testing with parallelism and coverage, check our xref:testing.adoc[testing guide].
```
good tip

=== Compiling

Use `nile compile` to compile contracts under the *contracts/* folder by default.

[,sh]
----
(env): nile compile
🤖 Compiling all Cairo contracts in the contracts directory
🔨 Compiling contracts/contract.cairo
✅ Done
----

TIP: For a full reference of Nile command options, check the xref:commands.adoc[CLI Reference] section.

=== Deploying an account

> . Add an environment variable with the Account Private Key to the *.env* file.

Author: I was trying to follow the following format rules:

- Bold for paths (filenames). Ex: **scripts/script.py**
- Code for code snippets. Ex: `nile init`

I personally like this more than using code for both, but no strong defense, if you think code is more appropriate I will update it.

> . Add an environment variable with the Account Private Key to the *.env* file.

Reviewer: yes, `backticks` should be used for both code and files/paths

+
[,sh]
----
# .env
ACCOUNT_1 = 0x1234
----
+
CAUTION: Don't use this Private Key in production environments!
+
. Use `nile counterfactual-address` to precompute the deployment address of the Account contract for the given signer.
+
[,sh]
----
(env): nile counterfactual-address ACCOUNT_1 --salt 123
0x057792c44b1e9e349ee549ed018f7225f927bd2b01b6277a1cb94da5fb9421ed
----
+
NOTE: Different salts compute to different addresses, even with the same Private Key.
+

> . In production, you will need to fund this address with Ether to pay for the deployment fees, which will be mandatory in testnets soon. There are multiple bridges and faucets, like link:https://goerli.starkgate.starknet.io/[Starkgate] being implemented at the moment to support the developer experience.

Reviewer: non production users need instructions too, or they might get stuck/drop in here. it's a good place to link the "Transfer funds from predeployed devnet account" sample script

Author: Added an entry for this.

+
. After funding the precomputed address, deploy the account using `nile setup`.
+
[,sh]
----
(env): nile setup ACCOUNT_1 --salt 123
----
+
[IMPORTANT]
====

> In StarkNet, users can deploy only previously declared contracts. {oz-account} is declared soon after every release in testnets and mainnet, but the local devnet needs a declaration for every instance, because the contract is not declared by default; however, you need an account to execute a declare transaction. To solve this egg-chicken problem, you can declare the OZ Account in devnet using predeployed accounts through scripts.

Reviewer: yes it is declared! devnet's accounts are OZ's. it's only sometimes when we develop the latest accounts that we're out of sync.

Author: Updated comment.

:pyenv: link:https://github.com/OpenZeppelin/cairo-contracts/blob/release-v0.4.0b/src/openzeppelin/access/ownable/library.cairo[pyenv]

= Overview

Nile is a CLI tool to develop or interact with StarkNet projects written in Cairo. It consists of different components for developing, compiling, testing, and deploying your smart contracts and dApps, providing a CLI for executing tasks, and a Runtime Environment (NRE) for scripting.

The package is designed to be extensible and very customizable by using plugins. This guide will take you through the installation of our recommended setup, but as we expect a lot of functionality coming from plugins, you are free to customize it.

TIP: You can get a list of plugins extending Nile functionality from the xref:plugins.adoc[Plugins] section.

== Requirements

> === GMP for fastecdsa

Reviewer: i believe this is an installation step so it probably belongs to quickstart. it makes me doubt this overview section should be a different one at all. what do you think about merging them?

tbh the gmp/fastecdsa issue is not a very sexy first impression for the docs

Author: Agree. I refactored it into just Overview.

✅ Done
----

TIP: For a full reference of Nile command options, check the xref:cli.adoc[CLI Reference] section.

=== Interacting with Starknet

==== Running a local node

By default, nile commands interacting directly with starknet nodes assume that you are using a local one if you don't specify the network using the `--network` option (default to `localhost`).

> Use `nile node` to run a local devnet node for local development.

Reviewer: ```suggestion
Use `nile node` to run a local devnet node for development.
```

[,sh]
----
(env): nile node
----

==== Deploying an account

> . Add an environment variable with the account's private pey to the `.env` file.

Reviewer: ```suggestion
. Add an environment variable with the account's private key to the `.env` file.
```

0x057792c44b1e9e349ee549ed018f7225f927bd2b01b6277a1cb94da5fb9421ed
----
+
NOTE: Different salts compute to different addresses, even with the same private key.
+
. In production, you will need to fund this address with Ether to pay for the deployment fees, which will be mandatory in testnets soon. There are multiple bridges and faucets, like link:https://goerli.starkgate.starknet.io/[Starkgate] being implemented at the moment to support the developer experience.
+

> For local development using devnet nodes, you can use predeployed accounts to fund your address by using xref:scripts.adoc#transfer_funds_from_predeployed_devnet_account[this script].

Reviewer: ```suggestion
For local development using a devnet node, you can use predeployed accounts to fund your address by using xref:scripts.adoc#transfer_funds_from_predeployed_devnet_account[this script].
```

> For local development using devnet nodes, you can use predeployed accounts to fund your address by using xref:scripts.adoc#transfer_funds_from_predeployed_devnet_account[this script].

Reviewer: ```suggestion
For local development using a devnet node, you can use predeployed accounts to fund your address by using xref:scripts.adoc#transfer_funds_from_predeployed_devnet_account[this script].
```

+
[IMPORTANT]
====

> In StarkNet, users can deploy only previously declared contracts. {oz-account} is declared soon after every release in devnet, testnets and mainnet, but sometimes there is an out-of-sync period between account update release and devnet release. Then you can declare the OZ Account using predeployed accounts through scripts.

Reviewer: ```suggestion
In StarkNet, users can deploy only previously declared contracts. {oz-account} is declared soon after every release in devnet, testnets and mainnet, but there could be an out-of-sync period between account update release and devnet release. Then you can declare the OZ Account using predeployed accounts through scripts.
```
